### PR TITLE
Add Clay.Threading and Clay.WaitAndRetry

### PR DESCRIPTION
**Pull request by grant-d**
Adds `Clay.Threading`, `Clay.WaitAndRetry` and moves `DistributedId` into `Clay.Primitives`

### CODE REVIEW

File: src/SourceCode.Clay.WaitAndRetry/ConcurrentRandom.cs [hunk 1]
```diff
@@ -0,0 +1,70 @@
...
+using System;
+using System.Diagnostics;
+
+namespace SourceCode.Clay.WaitAndRetry
+{
+    /// <summary>
+    /// A random number generator with a Uniform distribution that is thread-safe (via locking).
+    /// Can be instantiated with a custom <see cref="int"/> seed to make it emit deterministically.
+    /// </summary>
+    internal sealed class ConcurrentRandom
```
Comment (grant-d) [inline src/SourceCode.Clay.WaitAndRetry/ConcurrentRandom.cs]: I will move this into `Clay.Random` later, and make it `public`

File: src/SourceCode.Clay.Threading/CancellationTokenExtensions.cs [hunk 2]
```diff
@@ -0,0 +1,38 @@
...
+        /// and configured using the specified <paramref name="timeout"/>.
+        /// </summary>
+        /// <param name="cancellationToken">The <see cref="CancellationToken"/> to add a timeout to.</param>
+        public static CancellationTokenSource WithTimeout(this CancellationToken cancellationToken, TimeSpan timeout)
+        {
+            if (cancellationToken == CancellationToken.None)
+                return new CancellationTokenSource(timeout);
+
+            if (timeout == Timeout.InfiniteTimeSpan)
+                return new CancellationTokenSource();
```
Comment (jcdickinson) [inline src/SourceCode.Clay.Threading/CancellationTokenExtensions.cs]: `return cancellationToken`?

Comment (jcdickinson) [reply src/SourceCode.Clay.Threading/CancellationTokenExtensions.cs]: Or throw on infinite.

Comment (grant-d) [reply src/SourceCode.Clay.Threading/CancellationTokenExtensions.cs]: Fixed - Throw

**Pull request merged**